### DISCUSSION TITLE
Can tinygrad automatically split larger models across gpus while training?

### DISCUSSION THREAD

**Discussion by dhandhalyabhavik** [tinygrad/tinygrad#4696] [Q&A] [answered]
Wondering, if I want to train model similar to llama-70b from scratch on 2 gpus 24gb memory each. Will tinygrad automatically split the model for training?

I am talking about model parallelism not Data parallelism.

**Comment by chenyuxyz** [comment, marked as answer]
Sharding model is supported, we don't explicitly distinguish model parallelism from data parallelism https://github.com/tinygrad/tinygrad/blob/ce46a7e83ff2457f3c8a6b780ecf55ad2fa4a868/test/test_multitensor.py#L25

It's not automatic and you need to specify how you want to shard the model. This example shards 70B llama onto 6 GPU for serving, and you would need a lot more GPUs for training https://github.com/tinygrad/tinygrad/blob/ce46a7e83ff2457f3c8a6b780ecf55ad2fa4a868/examples/llama.py#L283

**Comment by dhandhalyabhavik** [comment]
Just want to know your thoughts on tinygrad shard vs pytorch + accelerate (huggingface)
Which one will make my experience writing model seamless while not worrying about data vs model parallelism.

**Reply by justyx404** [reply]
Probably `accelerate` since it is basically wrapped around `torch.distributed` and `deepspeed` which handles those details for you, all you need to provide is the number of gpus.

Personally I think tinygrad's approach is much more expressive and flexible. It also removes the need for additional API wrapper.
